itou/emails/models.py

```diff
@@ -0,0 +1,25 @@
+from django.contrib.postgres.fields import ArrayField, CIEmailField
+from django.contrib.postgres.indexes import GinIndex
+from django.db import models
+from django.utils import timezone
+
+
+class Email(models.Model):
+    to = ArrayField(CIEmailField(), verbose_name="à")
+    cc = ArrayField(CIEmailField(), default=list, verbose_name="cc")
+    bcc = ArrayField(CIEmailField(), default=list, verbose_name="cci")
+    subject = models.TextField(verbose_name="sujet")
+    body_text = models.TextField(verbose_name="message")
+    from_email = CIEmailField(verbose_name="de")
+    reply_to = ArrayField(CIEmailField(), default=list, verbose_name="répondre à")
```

Comment:
Pourquoi rajouter des `CIEmailField` ? On ne l'utilise que pour 2 champs jusque là.

Postgres décourage `citext` depuis pg12 et Django a vite emboité le pas car c'est deprecated depuis 4.2 donc faudra y passer dans 5.1 dans tous les cas…

Je suis quand même d'avis pour tous ces nouveaux champs d'utiliser un `EmailField` avec une collation case insensitive (non déterministe, on aura potentiellement des souci avec les lookups `contains` ceci dit, à tester).
Pour l'email de `User` et l'autre ça pourra être MAJ dans une follow up PR.

Reply:
François aura peut-être encore la source sous le coude (IIRC c'était un ticket dans le bug tracker de Django) mais pour le cas spécifique des emails le `citext` restais plus adapté que la collation.

Reply:
Cette remarque a mené à une grande discussion entre Léo et moi en privé.

TL;DR : Le plus simple est de rester sur `CIEmailField` (éventuellement via https://github.com/voiio/django-citext pour éviter le warning de dépréciation)

Les emails sont des valeurs restreintes (pas d’accents, pas vraiment de logique de texte), une collation est overkill. La collation pose problème pour créer un index et faire des recherches (il faut toujours préciser la collation de la recherche, sinon l’index n’est pas utilisé).

On pourrait imaginer un schéma bien plus simple, en utilisant `UPPER()` [1] dans l’index pour l’unicité et les recherches. Léo a expérimenté d’enregistrer un lookup personnalisé pour s’assurer qu’on fait bien les `icontains`, `iexact`, etc [3]. Ce qui pourrait être implémenté avec une sous-classe d’`EmailField` (e.g. `IndexedEmailField`) pour être sûr de toujours appliquer la transformation.

Premier point d’attention : `django-allauth` a décidé d’utiliser `LOWER` pour les emails [2], alors que Django utilise `UPPER` [1].

Blocage : on ne peut pas appliquer une fonction à toutes les valeurs d’un `ArrayField` pour construire l’index GIN. Donc pas possible de faire `UPPER(to)`[4].

Finalement, `citext` continuant à exister et permettant de faire ce dont nous avons besoin sur les emails, on garde `CIEmailField`.

Voir aussi la discussion sur le Google Group de django-developers : https://groups.google.com/g/django-developers/c/nDMnO98nexY/m/fyY2J-5gAgAJ

---

[1] django utilise `UPPER` : https://github.com/django/django/blob/f812b927a541fecc8ee445e1fd4dbe9d0540d523/django/db/backends/postgresql/base.py#L154-L167

[2] https://github.com/pennersr/django-allauth/blob/701ef63cfd4508af0faf0671ac89c87e0d52a801/allauth/account/utils.py#L448-L487

[3]
```python
In [1]: from itou.users.models import User

In [2]: from django.db.models import EmailField, lookups

In [3]: str(User.objects.only("email").filter(email__contains="leo").query)
Out[3]: 'SELECT "users_user"."id", "users_user"."email" FROM "users_user" WHERE "users_user"."email"::citext LIKE %leo%'

In [4]: EmailField.register_lookup(lookups.IContains, "contains")
Out[4]: django.db.models.lookups.IContains

In [5]: str(User.objects.only("email").filter(email__contains="leo").query)
Out[5]: 'SELECT "users_user"."id", "users_user"."email" FROM "users_user" WHERE UPPER("users_user"."email"::citext) LIKE UPPER(%leo%)'
```

[4] La méthode pour faire `UPPER` sur toutes les valeurs d’un ARRAY n’est pas utilisable comme expression dans un index.

```sql
itou=# CREATE INDEX ON emails_email USING GIN (UPPER(UNNEST("to")));
ERROR:  set-returning functions are not allowed in index expressions
```